src/NuGet.Clients/PackageManagement.UI/Services/NuGetLockService.cs

```diff
@@ -18,126 +17,49 @@ public sealed class NuGetLockService : INuGetLockService, IDisposable
     {
         private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
 
-        public bool IsLockHeld => _semaphore.CurrentCount == 0;
+        private readonly AsyncLocal<int> _lockCount = new AsyncLocal<int>();
 
-        public IDisposable AcquireLock()
-        {
-            _semaphore.Wait();
-            return new SemaphoreLockReleaser(_semaphore);
-        }
-
-        public IAsyncLockAwaitable AcquireLockAsync(CancellationToken token)
-        {
-            return new SemaphoreLockAwaiter(_semaphore, token);
-        }
-
-        public void Dispose()
-        {
-            _semaphore.Dispose();
-        }
+        public bool IsLockHeld => _semaphore.CurrentCount == 0;
 
-        private class SemaphoreLockReleaser : IDisposable
+        public async Task<T> ExecuteNuGetOperationAsync<T>(Func<Task<T>> action, CancellationToken token)
         {
-            private readonly SemaphoreSlim _semaphore;
-            private bool _isDisposed = false;
-
-            public SemaphoreLockReleaser(SemaphoreSlim semaphore)
+            if (_lockCount.Value == 0)
             {
-                _semaphore = semaphore;
-            }
-
-            public void Dispose()
-            {
-                Dispose(true);
-                GC.SuppressFinalize(this);
-            }
-
-            private void Dispose(bool _)
-            {
-                if (_isDisposed)
-                {
-                    return;
-                }
-
+                _lockCount.Value++;
```

Comment:
All the calls are being stacked one after another in that same async control flow so original call which acquired semaphore will only release it which means all the nested calls are already being completed. Which is why i didn't add more complexity to it.